src/GraphTopology.cpp

```diff
@@ -13,469 +13,301 @@
-    }
+    std::vector<int> indptr(num_faces + 1);
+    std::vector<int> indices(num_faces * 2);
+    std::vector<double> data(num_faces * 2, 1);
+
+    indptr[0] = 0;
 
-    // Insert edges to the faces shared between processors
-    for (int i = 0; i < Agg_Agg_o->num_rows; i++)
-        for (int j = Agg_Agg_o_i[i]; j < Agg_Agg_o_i[i + 1]; j++)
+    const auto& agg_indptr = agg_agg_diag.GetIndptr();
+    const auto& agg_indices = agg_agg_diag.GetIndices();
+    int rows = agg_agg_diag.Rows();
+    int count = 0;
+
+    for (int i = 0; i < rows; ++i)
+    {
+        for (int j = agg_indptr[i]; j < agg_indptr[i + 1]; ++j)
         {
-            sharedattr = Agg_shareattr_map[Agg_Agg_o_j[j]];
-            for (int k = agg_edge_i[i]; k < agg_edge_i[i + 1]; k++)
+            if (agg_indices[j] > i)
             {
-                edge = agg_edge_j[k];
-                if (edge_shareattr_i[edge + 1] > edge_shareattr_i[edge])
-                {
-                    edge_shareattr_loc =
-                        edge_shareattr_j[edge_shareattr_i[edge]];
-                    if (edge_shareattr_map[edge_shareattr_loc] == sharedattr)
-                        face_edge_j[face_edge_nnz++] = edge;
-                }
+                indices[count * 2] = i;
+                indices[count * 2 + 1] = agg_indices[j];
+
+                count++;
+
+                indptr[count] = count * 2;
             }
         }
-    double* face_edge_data = new double [face_edge_nnz];
-    std::fill_n(face_edge_data, face_edge_nnz, 1.0);
-    mfem::SparseMatrix face_edge_tmp(face_edge_i, face_edge_j, face_edge_data,
-                                     nfaces, nedges);
-    face_edge_.Swap(face_edge_tmp);
-
-    // TODO: face_bdratt can be built when counting boundary faces
-    if (edge_boundaryattr)
-    {
-        mfem::SparseMatrix face_bdr = smoothg::Mult(face_edge_, *edge_boundaryattr);
-        face_bdratt_.Swap(face_bdr);
     }
 
-    // Complete face to aggregate table
-    mfem::SparseMatrix face_agg_tmp(face_Agg_i, face_Agg_j,
-                                    face_Agg_data, nfaces, nAggs);
-    face_Agg_.Swap(face_agg_tmp);
+    assert(count == num_faces);
 
-    mfem::SparseMatrix agg_face_tmp = smoothg::Transpose(face_Agg_);
-    Agg_face_.Swap(agg_face_tmp);
+    return SparseMatrix(std::move(indptr), std::move(indices), std::move(data),
+                        num_faces, num_aggs);
+}
 
-    // Build face "dof-true dof-dof" table from local face_edge and
-    // the edge "dof-true dof-dof" table
-    GenerateOffsets(comm_, nfaces, face_start_);
+SparseMatrix GraphTopology::MakeFaceEdge(const ParMatrix& agg_agg,
+                                         const ParMatrix& edge_ext_agg,
+                                         const SparseMatrix& agg_edge_ext,
+                                         const SparseMatrix& face_edge_ext)
+{
+    const auto& agg_agg_diag = agg_agg.GetDiag();
```

Comment:
I think `agg_agg_diag` is not really needed other than defining `num_aggs`, which can be obtained by `agg_agg_offd.Rows()`.